src/inmanta/agent/agent.py

```diff
@@ -192,24 +252,136 @@ async def _execute(self, ctx: handler.HandlerContext, requires: dict[ResourceIdS
                 ctx.exception(str(e))
             except SkipResource as e:
                 ctx.set_status(const.ResourceState.skipped)
-                ctx.warning(msg="Resource %(resource_id)s was skipped: %(reason)s", resource_id=self.resource.id, reason=e.args)
+                ctx.warning(msg="Resource %(resource_id)s was skipped: %(reason)s", resource_id=resource.id, reason=e.args)
             except Exception as e:
                 ctx.set_status(const.ResourceState.failed)
                 ctx.exception(
                     "An error occurred during deployment of %(resource_id)s (exception: %(exception)s",
-                    resource_id=self.resource.id,
+                    resource_id=resource.id,
                     exception=repr(e),
                 )
         finally:
             if provider is not None:
                 provider.close()
 
+    async def _report_resource_status(
+        self,
+        env_id: uuid.UUID,
+        resource_id: Id,
+        resource_status: const.ResourceState,
+        resource_change: const.Change,
+        ctx: Optional[handler.HandlerContext] = None,
+    ) -> tuple[Optional[const.ResourceState], const.Change, dict[ResourceVersionIdStr, dict[str, AttributeStateChange]]]:
+        changes: dict[ResourceVersionIdStr, dict[str, AttributeStateChange]] = {}
+        if ctx:
+            changes = {resource_id.resource_version_str(): ctx.changes}
+            response = await self.client.resource_deploy_done(
+                tid=env_id,
+                rvid=resource_id.resource_version_str(),
+                action_id=ctx.action_id,
+                status=resource_status,
+                messages=ctx.logs,
+                changes=changes,
+                change=resource_change,
+            )
+        else:
+            response = await self.client.resource_deploy_done(
+                tid=env_id,
+                rvid=resource_id.resource_version_str(),
+                action_id=uuid.uuid4(),
+                status=resource_status,
+            )
+
+        if response.code != 200:
+            LOGGER.error("Resource status update failed %s", response.result)
+
+        return resource_status, resource_change, changes
+
+    async def execute(
+        self,
+        ctx: handler.HandlerLogger,
+        gid: uuid.UUID,
+        resource_id: Id,
+        resource: dict[str, object],
+        env_id: uuid.UUID,
+        agent: "AgentInstance",
+    ) -> tuple[Optional[const.ResourceState], const.Change, dict[ResourceVersionIdStr, dict[str, AttributeStateChange]]]:
+        ctx.debug("Start deploy %(deploy_id)s of resource %(resource_id)s", deploy_id=gid, resource_id=resource_id)
```

Comment:
I would move this to after deserialization so we can use the proper handler context.